userspace/libsinsp/test/user.ut.cpp

```diff
@@ -0,0 +1,166 @@
+
+*/
+
+#include <fstream>
+#include <gtest/gtest.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "sinsp_with_test_input.h"
+#include "user.h"
+
+using namespace libsinsp;
+
+class usergroup_manager_test : public sinsp_with_test_input
+{
+	// for gtest filtering convenience,
+	// add something when needed
+};
+
+TEST_F(usergroup_manager_test, add_rm)
+{
+	std::string container_id{""};
+
+	sinsp_usergroup_manager mgr(&m_inspector);
+	// no data so far
+	ASSERT_EQ(mgr.get_user(container_id, 0), nullptr);
+	ASSERT_EQ(mgr.get_group(container_id, 0), nullptr);
+	ASSERT_EQ(mgr.get_userlist(container_id), nullptr);
+	ASSERT_EQ(mgr.get_grouplist(container_id), nullptr);
+
+	// user
+	mgr.add_user(container_id, 0, 0, "test", "/test", "/bin/test");
+	auto* user = mgr.get_user(container_id, 0);
+	ASSERT_NE(user, nullptr);
+	ASSERT_EQ(user->uid, 0);
+	ASSERT_EQ(user->gid, 0);
+	ASSERT_EQ(std::string(user->name), "test");
+	ASSERT_EQ(std::string(user->homedir), "/test");
+	ASSERT_EQ(std::string(user->shell), "/bin/test");
+
+	auto* userlist = mgr.get_userlist(container_id);
+	{
+		auto it = userlist->find(0);
+		ASSERT_NE(it, userlist->end());
+		ASSERT_EQ(&(it->second), user);
+	}
+
+	// group
+	mgr.add_group(container_id, 0, "test");
+	auto* group = mgr.get_group(container_id, 0);
+	ASSERT_NE(group, nullptr);
+	ASSERT_EQ(group->gid, 0);
+	ASSERT_EQ(std::string(group->name), "test");
+
+	auto* grouplist = mgr.get_grouplist(container_id);
+	{
+		auto it = grouplist->find(0);
+		ASSERT_NE(it, grouplist->end());
+		ASSERT_EQ(&(it->second), group);
+	}
+
+	// rm
+	mgr.rm_user(container_id, 0);
+	ASSERT_EQ(mgr.get_user(container_id, 0), nullptr);
+	mgr.rm_group(container_id, 0);
+	ASSERT_EQ(mgr.get_group(container_id, 0), nullptr);
+}
+
+TEST_F(usergroup_manager_test, system_lookup)
+{
+	std::string container_id{""};
+
+	sinsp_usergroup_manager mgr(&m_inspector);
+
+	mgr.add_user(container_id, 0, 0, nullptr, nullptr, nullptr);
+	auto* user = mgr.get_user(container_id, 0);
+	ASSERT_NE(user, nullptr);
+	ASSERT_EQ(user->uid, 0);
+	ASSERT_EQ(user->gid, 0);
+	ASSERT_EQ(std::string(user->name), "root");
+	ASSERT_EQ(std::string(user->homedir), "/root");
+	ASSERT_EQ(std::string(user->shell).empty(), false);
+
+	mgr.add_group(container_id, 0, nullptr);
+	auto* group = mgr.get_group(container_id, 0);
+	ASSERT_NE(group, nullptr);
+	ASSERT_EQ(group->gid, 0);
+	ASSERT_EQ(std::string(group->name), "root");
+}
+
+#if defined(HAVE_PWD_H) || defined(HAVE_GRP_H)
+class usergroup_manager_host_root_test : public sinsp_with_test_input
+{
+protected:
+	void SetUp() override
+	{
+		char pwd_buf[SCAP_MAX_PATH_SIZE];
+		auto pwd = getcwd(pwd_buf, SCAP_MAX_PATH_SIZE);
+		ASSERT_NE(pwd, nullptr);
+		m_host_root = pwd_buf;
+		m_host_root += "/host";
+
+		ASSERT_EQ(mkdir(m_host_root.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH), 0);
+		std::string etc = m_host_root + "/etc";
+		ASSERT_EQ(mkdir(etc.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH), 0);
```

Comment:
I want to state that in my experience, tests that rely on the underlying system are usually flaky and finicky. I feel it'd be better if we could directly mock the calls to the system, but since gtest doesn't allow for that in an easy way... Here's hoping 🤞🏻

Reply:
I feel bad too, but I don't know how else I could do that. So :crossed_fingers: :crossed_fingers: